[improve](hive) Refactor csv reader

suxiaogang223 opened the pull request:
### What problem does this PR solve?

- [x] Impl TextReader for reading hive text table.
- [x] Cleanup CsvReader
- [x] Add cases to test csv and text format

### Release note

behavior changes:
1. Don't parse "\\N" or other string as null for hive OpenCsv table
In Hive's OpenCsv table, there is no definition for the representation of null values. This behavior is different from the behavior of importing CSV tables internally in Doris.
2. Fix the bug when reading complex type for hive OpenCsv table
The current code incorrectly uses the hive text format to parse the complex types of Hive OpenCsv tables. This PR fixes this behavior and uses the json format for parsing.


### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

suxiaogang223 committed:
use double quote to escape quote

escape quote for csv

intro escape_string_for_csv and deserialize_one_cell_from_csv

fix csv_reader

add regression

flush out

fix regression data

add be ut for EncloseCsvLineReaderContext and PlainTextLineReaderCtx

fix ut

fix ut

fix ut

test custom enclose

reset

fix ut : reset ctx after read line

remove unuseful codes

suxiaogang223 committed:
intro TextReader

remove unused field in csv_reader

cleanup text_reader

test TextReader

fix build

fix TextReader

clean csv_reader

intro TFileFormatType::FORMAT_TEXT

fix hive text

cleanup csv_reader

support load task for text_reader

csv format don't consider null value

clean code about TTextSerdeType

for csv: col idx is out of range, fill with empty string.

fix build

fix

consider nullable for load

consider null for load csv

add case

set null format with empty string to make csv reader not use "\\N" to represent null

fix pipeline

impl TextFileFormatProperties

fix tvf

fix build

hello-stephen commented:
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

hello-stephen commented:
# BE Regression P0 && UT Coverage Report
Increment line coverage `4.55% (13/286)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_5650a1d1a11b09aa9f804cf810d60aec61718c1f_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_5650a1d1a11b09aa9f804cf810d60aec61718c1f_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage |                    |
| Line Coverage     |                    |
| Region Coverage   |                    |
| Branch Coverage   |                    |

suxiaogang223 commented:
run external

suxiaogang223 commented:
run external

suxiaogang223 commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `4.58% (13/284)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/52a7949fc04539b6896be4dc0a35c46db90b026a_52a7949fc04539b6896be4dc0a35c46db90b026a/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/52a7949fc04539b6896be4dc0a35c46db90b026a_52a7949fc04539b6896be4dc0a35c46db90b026a/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 54.44% (14766/27122) |
| Line Coverage     | 43.52% (129207/296900) |
| Region Coverage   | 42.22% (65988/156296) |
| Branch Coverage   | 36.75% (33245/90470) |

hello-stephen commented:
# BE Regression P0 && UT Coverage Report
Increment line coverage `4.55% (13/286)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_52a7949fc04539b6896be4dc0a35c46db90b026a_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_52a7949fc04539b6896be4dc0a35c46db90b026a_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 55.52% (14782/26626) |
| Line Coverage     | 45.14% (133795/296416) |
| Region Coverage   | 42.07% (76893/182753) |
| Branch Coverage   | 36.08% (37174/103044) |

suxiaogang223 commented:
run buildall

hello-stephen commented:
# BE Regression P0 && UT Coverage Report
Increment line coverage `18.31% (52/284)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_7a772c870025b13d35a08d10543510c94f52d3bf_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_7a772c870025b13d35a08d10543510c94f52d3bf_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 78.02% (20660/26480) |
| Line Coverage     | 71.57% (210965/294764) |
| Region Coverage   | 69.76% (126774/181718) |
| Branch Coverage   | 63.01% (64440/102268) |

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 committed:
add cases

add decimal_separators

add headers

add multidelim

add unescaped_quotes

add cases

trim tailing space

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `3.23% (13/402)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/aa941a4bdc89d862da03848b2752576437517703_aa941a4bdc89d862da03848b2752576437517703/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/aa941a4bdc89d862da03848b2752576437517703_aa941a4bdc89d862da03848b2752576437517703/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 55.07% (14918/27087) |
| Line Coverage     | 44.18% (131109/296769) |
| Region Coverage   | 42.85% (66810/155921) |
| Branch Coverage   | 37.41% (33693/90076) |

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run external

suxiaogang223 commented:
run external

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run external

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `85.82% (345/402)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_07236b6a14c6d794c68952ce96264e4193e00838_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_07236b6a14c6d794c68952ce96264e4193e00838_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 81.02% (21273/26257) |
| Line Coverage     | 74.94% (220091/293696) |
| Region Coverage   | 73.24% (132435/180819) |
| Branch Coverage   | 66.59% (67628/101564) |

morningman commented on be/src/vec/data_types/serde/data_type_string_serde.h:
```diff
@@ -64,6 +64,31 @@ inline void escape_string(const char* src, size_t& len, char escape_char) {
 }
 
+// specially escape quote with double quote
+inline void escape_string_for_csv(const char* src, size_t& len, char escape_char, char quote_char) {
```
better use pointer for out parameter

morningman commented on be/src/vec/exec/format/csv/csv_reader.cpp:
```diff
@@ -395,14 +328,10 @@ Status CsvReader::init_reader(bool is_load) {
     if (_enclose == 0) {
         text_line_reader_ctx = std::make_shared<PlainTextLineReaderCtx>(
```
How about unify the naming of `PlainTextLineReaderCtx` and `EncloseCsvLineReaderContext`?
eg:
- PlainCsvLineReaderCtx
- EncloseCsvLineReaderCtx

morningman commented on be/src/vec/exec/format/csv/csv_reader.cpp:
```diff
@@ -395,14 +328,10 @@ Status CsvReader::init_reader(bool is_load) {
-            _fields_splitter = std::make_unique<PlainCsvTextFieldSplitter>(
-                    _trim_tailing_spaces, false, _value_separator, _value_separator_length, -1);
-        }
+        // TODO: make sure what table this splitter is used for
```
What is this TODO mean?

suxiaogang223 replied:
Just forget to rm this

morningman commented on be/src/vec/exec/format/csv/csv_reader.cpp:
```diff
@@ -649,11 +577,10 @@ Status CsvReader::_fill_dest_columns(const Slice& line, Block* block,
+        // col idx is out of range, fill with null format
+        auto value = col_idx < _split_values.size()
+                             ? _split_values[col_idx]
+                             : Slice(_options.null_format, _options.null_len);
```
the `_options.null_len` may be zero?
And as you said, there is no definition for `null` value in csv, so how do we define `null` here?

suxiaogang223 replied:
In the process of doris load csv, "\\N" is hard-coded as a null value, but there is no null value in the opencsv format of hive, and empty strings are used to fill non-existent columns. When reading the hive csv table here, null_format will be an empty string, and null_len will be 0

morningman commented on be/src/vec/exec/format/text/text_reader.cpp:
```diff
@@ -0,0 +1,466 @@
+namespace doris::vectorized {
+#include "common/compile_check_begin.h"
+
+void HiveCsvTextFieldSplitter::do_split(const Slice& line, std::vector<Slice>* splitted_values) {
```
```suggestion
void HiveTextFieldSplitter::do_split(const Slice& line, std::vector<Slice>* splitted_values) {
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/LoadScanProvider.java:
```diff
@@ -97,7 +96,8 @@ public FileLoadScanNode.ParamCreateContext createContext(Analyzer analyzer) thro
         if (fileGroupInfo.getFileGroup().getFileFormat() != null
                 && fileGroupInfo.getFileGroup().getFileFormat().equals("hive_text")) {
-            params.setTextSerdeType(TTextSerdeType.HIVE_TEXT_SERDE);
+            // TODO: set correct format type for load job
```
How to resolve this TODO?

suxiaogang223 replied:
fixed

suxiaogang223 committed:
fix test_open_csv_serde

fix

support stream_load hive_text

add regression for stream load hive_text

fix

add ut for escape_string

fix be ut

fix regression

fix load_util.cpp

fix build

fix bt ut

refactor csv and text reader

fix be ut

fix

fix be core

fix build

fix external

fix merge

suxiaogang223 commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `5.53% (23/416)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/e3ab06d3022d325590a75717a047bb43aad698c9_e3ab06d3022d325590a75717a047bb43aad698c9/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/e3ab06d3022d325590a75717a047bb43aad698c9_e3ab06d3022d325590a75717a047bb43aad698c9/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 55.73% (14919/26770) |
| Line Coverage     | 44.50% (131539/295584) |
| Region Coverage   | 43.41% (66285/152707) |
| Branch Coverage   | 38.09% (33854/88882) |

suxiaogang223 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/f75ee7ab79d79a9234b6f57fbb1812e137a1d560_f75ee7ab79d79a9234b6f57fbb1812e137a1d560_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/f75ee7ab79d79a9234b6f57fbb1812e137a1d560_f75ee7ab79d79a9234b6f57fbb1812e137a1d560_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.20% (1104/1327) |
| Line Coverage     | 66.12% (18603/28136) |
| Region Coverage   | 65.68% (9260/14098) |
| Branch Coverage   | 55.40% (4976/8982) |

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `77.88% (345/443)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_ce0d08255e5ac9042d3c5272d45360605f1ea8d5_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_ce0d08255e5ac9042d3c5272d45360605f1ea8d5_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.50% (20893/26282) |
| Line Coverage     | 72.68% (215005/295814) |
| Region Coverage   | 70.82% (126520/178639) |
| Branch Coverage   | 64.48% (65461/101518) |

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run external

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/43e7251cf041c9b92b6e83e38b7fbc7f18294c78_43e7251cf041c9b92b6e83e38b7fbc7f18294c78_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/43e7251cf041c9b92b6e83e38b7fbc7f18294c78_43e7251cf041c9b92b6e83e38b7fbc7f18294c78_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.25% (1113/1337) |
| Line Coverage     | 66.15% (18669/28223) |
| Region Coverage   | 65.81% (9264/14077) |
| Branch Coverage   | 55.60% (4984/8964) |

suxiaogang223 commented:
run buildall

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `23.38% (72/308)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/2253603deb065239177f2befdc0eac6184adfc0a_2253603deb065239177f2befdc0eac6184adfc0a/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/2253603deb065239177f2befdc0eac6184adfc0a_2253603deb065239177f2befdc0eac6184adfc0a/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 56.06% (14943/26653) |
| Line Coverage     | 44.87% (132917/296227) |
| Region Coverage   | 43.96% (66835/152043) |
| Branch Coverage   | 38.60% (34300/88862) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `56.35% (173/307)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_2253603deb065239177f2befdc0eac6184adfc0a_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_2253603deb065239177f2befdc0eac6184adfc0a_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 60.43% (15853/26235) |
| Line Coverage     | 49.96% (148002/296213) |
| Region Coverage   | 47.21% (84305/178573) |
| Branch Coverage   | 40.80% (41409/101496) |

suxiaogang223 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/9dfcc92756376ccf764ec9b3ce62f40a5c6f3540_9dfcc92756376ccf764ec9b3ce62f40a5c6f3540_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/9dfcc92756376ccf764ec9b3ce62f40a5c6f3540_9dfcc92756376ccf764ec9b3ce62f40a5c6f3540_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.25% (1113/1337) |
| Line Coverage     | 66.19% (18681/28224) |
| Region Coverage   | 65.89% (9276/14079) |
| Branch Coverage   | 55.63% (4988/8966) |

suxiaogang223 commented:
run buildall

suxiaogang223 commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/3a662dc35fb3163b0aebab385d2348d8bb2b64d7_3a662dc35fb3163b0aebab385d2348d8bb2b64d7_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/3a662dc35fb3163b0aebab385d2348d8bb2b64d7_3a662dc35fb3163b0aebab385d2348d8bb2b64d7_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 83.31% (1118/1342) |
| Line Coverage     | 66.76% (19136/28663) |
| Region Coverage   | 66.40% (9469/14261) |
| Branch Coverage   | 56.35% (5131/9106) |

starocean999 commented on fe/fe-core/src/main/java/org/apache/doris/datasource/LoadScanProvider.java:
```diff
@@ -95,10 +94,7 @@ public FileLoadScanNode.ParamCreateContext createContext(Analyzer analyzer) thro
-            params.setTextSerdeType(TTextSerdeType.HIVE_TEXT_SERDE);
-        }
-        params.setProperties(fileGroupInfo.getBrokerDesc().getBackendConfigProperties());
+        params.setProperties(fileGroupInfo.getBrokerDesc().getProperties());
```
looks like createContext should not be called in latest master, could you please show an example that how this function is called?

suxiaogang223 committed:
fix

suxiaogang223 committed:
fix be ut

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  47.06% (24/51)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_0587de7ce408c6cdb9b6a233bb918066405641ad/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_0587de7ce408c6cdb9b6a233bb918066405641ad/fe_report/index.html)

suxiaogang223 committed:
fix p0 coredump

suxiaogang223 committed:
fix

suxiaogang223 committed:
fix p0

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `  47.06% (24/51)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_5d3d0109fc32ee4e564be68bd5c289a755438b30/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_5d3d0109fc32ee4e564be68bd5c289a755438b30/fe_report/index.html)

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `88.10% (237/269)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/50379_5d3d0109fc32ee4e564be68bd5c289a755438b30_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/50379_5d3d0109fc32ee4e564be68bd5c289a755438b30_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.57% (20999/26390) |
| Line Coverage     | 72.64% (216853/298536) |
| Region Coverage   | 70.87% (127721/180212) |
| Branch Coverage   | 64.61% (66150/102388) |
